Framework/Geometry/src/Objects/CSGObject.cpp

```diff
@@ -115,11 +115,11 @@ void CSGObject::setMaterial(const Kernel::Material &material) {
 /**
  * @return The Material that the object is composed from
  */
-const Kernel::Material CSGObject::material() const {
-  if (m_material)
-    return *m_material;
-  else
-    return Material();
+const Kernel::Material &CSGObject::material() const {
+  if (!m_material) {
+    m_material = Kernel::make_unique<Material>();
```

Comment:
These old files are actually less scary to modify than new ones since they are covered by lots of unit, doc and system tests at different levels.

The original intention regarding the `material()` function is to clearly return a `const` reference. However, this would return a dangling reference the way `CSGObject::material()` was implemented previously, thus somebody removed the `&` without a second thought.

I believe the reason `CSGObject` uses a `unique_ptr` is to save some memory as there can be a hundred thousand `CSGObject` detectors without material and a single sample shape with one. For some reason, `MeshObject` and `MeshObject2D` went a different route.

Anyway, I think this change gives us the originally intended behavior --- returning the material object as `const` reference.

Reply:
I was speaking with @gemmaguest and since it is a const method, should it really be making a change to the object?

I agree that it could be the expected behavior, however since it is empty by default surely that may have been for a reason, so should it not just be set then? Thus negating the point of this discussion.

Reply:
From a user's point of view (looking at this class as a black box):
* If no material is set this method returns a `const` reference to an empty material. No modifications to the `CSGObject` object.
* If material is set this method returns a `const` reference to the material. No modifications to the `CSGObject` object.

This is exactly the previous behavior with the exception that now we return a `const` reference to `Material`, not a copy.

Reply:
I see your point!